src/main/java/com/carhouse/controller/CarSaleController.java

```diff
@@ -90,16 +90,91 @@ public String addCarSale(final Model model) {
     /**
      * Submit add car sale advertisement.
      * Take list of selected car feature's id as request param
-     * Return home page
+     * Return page with list car sales
      *
      * @param carSale     object is used to get entered data.
      * @param featureList the feature list
      * @return view string
      */
     @PostMapping("/carSale/add")
     public String addCarSaleSubmit(@ModelAttribute final CarSale carSale,
-                                   @RequestParam(value = "carFeatureList") final int[] featureList) {
+                                   @RequestParam(required = false, value = "carFeatureList") final int[] featureList) {
         carSaleProvider.addCarSale(carSale, featureList);
-        return "redirect:/homePage";
+        return "redirect:/carSale";
+    }
+
+    /**
+     * Return page to update new car sale advertisement.
+     * Create list of selected car feature's id to select car feature
+     * Gets the request url to return to the same page
+     *
+     * @param carSaleId  the car sale id
+     * @param requestUrl the request url
+     * @param model      model
+     * @return view string
+     */
+    @GetMapping("/carSale/{carSaleId}/update")
+    public String updateCarSale(@PathVariable final int carSaleId,
+                                @RequestParam("requestUrl") final String requestUrl, final Model model) {
+        CarSale carSale = carSaleProvider.getCarSale(carSaleId);
+        model.addAttribute("carSale", carSale);
+        model.addAttribute("requestUrl", requestUrl);
+        model.addAttribute("selectedCarFeatures",
+                createSelectedCarFeatureList(carSale.getCar().getCarFeatureList()));
+        model.addAttribute("listCarMakes", carMakeProvider.getCarMakes());
+        model.addAttribute("listFuelTypes", fuelTypeProvider.getFuelTypes());
+        model.addAttribute("listCarFeatures", carFeatureProvider.getCarFeatures());
+        return "updateCarSale";
+    }
+
+    /**
+     * Submit update car sale advertisement.
+     * Take list of selected car feature's id as request param
+     * Return page with list car sales
+     *
+     * @param carSale     the car sale
+     * @param requestUrl  the request url
+     * @param featureList the feature list
+     * @return the string
+     */
+    @PostMapping("/carSale/update")
+    public String updateCarSaleSubmit(
+            @ModelAttribute final CarSale carSale,
+            @RequestParam("requestUrl") final String requestUrl,
+            @RequestParam(required = false, value = "carFeatureList") final int[] featureList) {
+        carSaleProvider.updateCarSale(carSale, featureList);
+        return "redirect:" + requestUrl;
+    }
+
+    /**
+     * Delete car sale string.
+     * Gets the request url to return to the same page
+     *
+     * @param carSaleId  the car sale id
+     * @param requestUrl the request url
+     * @return the string
+     */
+    @GetMapping("/carSale/{carSaleId}/delete")
```

Comment:
get for delete operation?

Reply:
What is there wrong?
It is a web controller and as I know It uses only `get` and `post` mappings
so I chose `get` for delete operations